### ISSUE TITLE
update mutation with new cli does not work???

### ISSUE THREAD

**Issue by rush86999** [aws-amplify/amplify-cli#3726]
>**Note**: If your issue/bug is regarding the AWS Amplify Console service, please log it in the 
[Amplify Console GitHub Issue Tracker](https://github.com/aws-amplify/amplify-console/issues)

**Describe the bug**
Create a new project with schema. I am unable to update any of the values with generated mutation graphql mutations with cli. I make sure the required input values are included in the inputs to update the non-required values when I run my mutations but no change. I have also tried the appsync console with no avail. I have erased and tried a whole new environment in the cli but the bug keeps repeating. I am using 4.16.1. I have Merge enabled. The backend schema has

```
_version: Int!
  _deleted: Boolean
  _lastChangedAt: AWSTimestamp!
```
included.  I also have iam provider and cognito user pools for auth rules. I am trying to update Profile table for my webapp.

here the VTL template
```
## [Start] Determine request authentication mode **
#if( $util.isNullOrEmpty($authMode) && !$util.isNull($ctx.identity) && !$util.isNull($ctx.identity.sub) && !$util.isNull($ctx.identity.issuer) && !$util.isNull($ctx.identity.username) && !$util.isNull($ctx.identity.claims) && !$util.isNull($ctx.identity.sourceIp) && !$util.isNull($ctx.identity.defaultAuthStrategy) )
  #set( $authMode = "userPools" )
#end
## [End] Determine request authentication mode **
## [Start] Check authMode and execute owner/group checks **
#if( $authMode == "userPools" )
  ## [Start] Static Group Authorization Checks **
  #set($isStaticGroupAuthorized = $util.defaultIfNull(
            $isStaticGroupAuthorized, false))
  ## Authorization rule: { allow: groups, groups: ["Users"], groupClaim: "cognito:groups" } **
  #set( $userGroups = $util.defaultIfNull($ctx.identity.claims.get("cognito:groups"), []) )
  #set( $allowedGroups = ["Users"] )
  #foreach( $userGroup in $userGroups )
    #if( $allowedGroups.contains($userGroup) )
      #set( $isStaticGroupAuthorized = true )
      #break
    #end
  #end
  ## [End] Static Group Authorization Checks **


  #if( ! $isStaticGroupAuthorized )
    ## No dynamic group authorization rules **


    ## No owner authorization rules **


    ## [Start] Collect Auth Condition **
    #set( $authCondition = $util.defaultIfNull($authCondition, {
  "expression": "",
  "expressionNames": {},
  "expressionValues": {}
}) )
    #set( $totalAuthExpression = "" )
    ## Add dynamic group auth conditions if they exist **
    #if( $groupAuthExpressions )
      #foreach( $authExpr in $groupAuthExpressions )
        #set( $totalAuthExpression = "$totalAuthExpression $authExpr" )
        #if( $foreach.hasNext )
          #set( $totalAuthExpression = "$totalAuthExpression OR" )
        #end
      #end
    #end
    #if( $groupAuthExpressionNames )
      $util.qr($authCondition.expressionNames.putAll($groupAuthExpressionNames))
    #end
    #if( $groupAuthExpressionValues )
      $util.qr($authCondition.expressionValues.putAll($groupAuthExpressionValues))
    #end
    ## Add owner auth conditions if they exist **
    #if( $totalAuthExpression != "" && $ownerAuthExpressions && $ownerAuthExpressions.size() > 0 )
      #set( $totalAuthExpression = "$totalAuthExpression OR" )
    #end
    #if( $ownerAuthExpressions )
      #foreach( $authExpr in $ownerAuthExpressions )
        #set( $totalAuthExpression = "$totalAuthExpression $authExpr" )
        #if( $foreach.hasNext )
          #set( $totalAuthExpression = "$totalAuthExpression OR" )
        #end
      #end
    #end
    #if( $ownerAuthExpressionNames )
      $util.qr($authCondition.expressionNames.putAll($ownerAuthExpressionNames))
    #end
    #if( $ownerAuthExpressionValues )
      $util.qr($authCondition.expressionValues.putAll($ownerAuthExpressionValues))
    #end
    ## Set final expression if it has changed. **
    #if( $totalAuthExpression != "" )
      #if( $util.isNullOrEmpty($authCondition.expression) )
        #set( $authCondition.expression = "($totalAuthExpression)" )
      #else
        #set( $authCondition.expression = "$authCondition.expression AND ($totalAuthExpression)" )
      #end
    #end
    ## [End] Collect Auth Condition **
  #end


  ## [Start] Throw if unauthorized **
  #if( !($isStaticGroupAuthorized == true || ($totalAuthExpression != "")) )
    $util.unauthorized()
  #end
  ## [End] Throw if unauthorized **
#end
## [End] Check authMode and execute owner/group checks **

#if( $authCondition && $authCondition.expression != "" )
  #set( $condition = $authCondition )
  #if( $modelObjectKey )
    #foreach( $entry in $modelObjectKey.entrySet() )
      $util.qr($condition.put("expression", "$condition.expression AND attribute_exists(#keyCondition$velocityCount)"))
      $util.qr($condition.expressionNames.put("#keyCondition$velocityCount", "$entry.key"))
    #end
  #else
    $util.qr($condition.put("expression", "$condition.expression AND attribute_exists(#id)"))
    $util.qr($condition.expressionNames.put("#id", "id"))
  #end
#else
  #if( $modelObjectKey )
    #set( $condition = {
  "expression": "",
  "expressionNames": {},
  "expressionValues": {}
} )
    #foreach( $entry in $modelObjectKey.entrySet() )
      #if( $velocityCount == 1 )
        $util.qr($condition.put("expression", "attribute_exists(#keyCondition$velocityCount)"))
      #else
        $util.qr($condition.put("expression", "$condition.expression AND attribute_exists(#keyCondition$velocityCount)"))
      #end
      $util.qr($condition.expressionNames.put("#keyCondition$velocityCount", "$entry.key"))
    #end
  #else
    #set( $condition = {
  "expression": "attribute_exists(#id)",
  "expressionNames": {
      "#id": "id"
  },
  "expressionValues": {}
} )
  #end
#end
## Automatically set the updatedAt timestamp. **
$util.qr($context.args.input.put("updatedAt", $util.defaultIfNull($ctx.args.input.updatedAt, $util.time.nowISO8601())))
$util.qr($context.args.input.put("__typename", "Profile"))
## Update condition if type is @versioned **
#if( $versionedCondition )
  $util.qr($condition.put("expression", "($condition.expression) AND $versionedCondition.expression"))
  $util.qr($condition.expressionNames.putAll($versionedCondition.expressionNames))
  $util.qr($condition.expressionValues.putAll($versionedCondition.expressionValues))
#end
#if( $context.args.condition )
  #set( $conditionFilterExpressions = $util.parseJson($util.transform.toDynamoDBConditionExpression($context.args.condition)) )
  $util.qr($condition.put("expression", "($condition.expression) AND $conditionFilterExpressions.expression"))
  $util.qr($condition.expressionNames.putAll($conditionFilterExpressions.expressionNames))
  $util.qr($condition.expressionValues.putAll($conditionFilterExpressions.expressionValues))
#end
#if( $condition.expressionValues && $condition.expressionValues.size() == 0 )
  #set( $condition = {
  "expression": $condition.expression,
  "expressionNames": $condition.expressionNames
} )
#end
#set( $expNames = {} )
#set( $expValues = {} )
#set( $expSet = {} )
#set( $expAdd = {} )
#set( $expRemove = [] )
#if( $modelObjectKey )
  #set( $keyFields = [] )
  #foreach( $entry in $modelObjectKey.entrySet() )
    $util.qr($keyFields.add("$entry.key"))
  #end
#else
  #set( $keyFields = ["id", "_version", "_deleted", "_lastChangedAt"] )
#end
#foreach( $entry in $util.map.copyAndRemoveAllKeys($context.args.input, $keyFields).entrySet() )
  #if( !$util.isNull($dynamodbNameOverrideMap) && $dynamodbNameOverrideMap.containsKey("$entry.key") )
    #set( $entryKeyAttributeName = $dynamodbNameOverrideMap.get("$entry.key") )
  #else
    #set( $entryKeyAttributeName = $entry.key )
  #end
  #if( $util.isNull($entry.value) )
    #set( $discard = $expRemove.add("#$entryKeyAttributeName") )
    $util.qr($expNames.put("#$entryKeyAttributeName", "$entry.key"))
  #else
    $util.qr($expSet.put("#$entryKeyAttributeName", ":$entryKeyAttributeName"))
    $util.qr($expNames.put("#$entryKeyAttributeName", "$entry.key"))
    $util.qr($expValues.put(":$entryKeyAttributeName", $util.dynamodb.toDynamoDB($entry.value)))
  #end
#end
#set( $expression = "" )
#if( !$expSet.isEmpty() )
  #set( $expression = "SET" )
  #foreach( $entry in $expSet.entrySet() )
    #set( $expression = "$expression $entry.key = $entry.value" )
    #if( $foreach.hasNext() )
      #set( $expression = "$expression," )
    #end
  #end
#end
#if( !$expAdd.isEmpty() )
  #set( $expression = "$expression ADD" )
  #foreach( $entry in $expAdd.entrySet() )
    #set( $expression = "$expression $entry.key $entry.value" )
    #if( $foreach.hasNext() )
      #set( $expression = "$expression," )
    #end
  #end
#end
#if( !$expRemove.isEmpty() )
  #set( $expression = "$expression REMOVE" )
  #foreach( $entry in $expRemove )
    #set( $expression = "$expression $entry" )
    #if( $foreach.hasNext() )
      #set( $expression = "$expression," )
    #end
  #end
#end
#set( $update = {} )
$util.qr($update.put("expression", "$expression"))
#if( !$expNames.isEmpty() )
  $util.qr($update.put("expressionNames", $expNames))
#end
#if( !$expValues.isEmpty() )
  $util.qr($update.put("expressionValues", $expValues))
#end
{
  "version": "2018-05-29",
  "operation": "UpdateItem",
  "key": #if( $modelObjectKey ) $util.toJson($modelObjectKey) #else {
  "id": {
      "S": $util.toJson($context.args.input.id)
  }
} #end,
  "update": $util.toJson($update),
  "condition": $util.toJson($condition),
  "_version": $util.defaultIfNull($ctx.args.input["_version"], "0")
}
```

here's schema for table
```
type Profile
@model
@auth(rules: [
  { allow: public, provider: iam, operations: [read] },
  {allow: groups, groups: ["public"], operations: [read]},
  {allow: groups, groups: ["Users"]},
])
{
  id: ID!
  userId: ID!
  name: String!
  aboutMe: String
  profilePicture: String
  following: [UserHost] @connection(keyName: "byUser", fields: ["id"])
  followingCount: String
  user: User @connection(fields: ["userId"])
  stats: Stats @connection(fields: ["userId"])
}
```

here's my react code:

```
const updateUserProfileForm = async (values) => {
    try {
      const {
        files,
        aboutMe,
      } = values;

      //  upload picture
      const filteredFiles = files.filter(file => file !== null);
      console.log(values, ' this is the values object');
      console.log(files, ' this is files');
      if (filteredFiles
        && filteredFiles[0]
        && filteredFiles[0].name) await uploadPicture(filteredFiles[0]);

      const params = (filteredFiles && filteredFiles[0] && filteredFiles[0].name)
        ? {
          input: {
            id: profileId,
            aboutMe,
            profilePicture: `${credId}/${profileId}/${escapeUnsafe(filteredFiles[0].name)}`,
            userId: userProfile.userId,
            name: userProfile.name,
          },
        }
        : {
          input: {
            id: profileId,
            aboutMe,
            userId: userProfile.userId,
            name: userProfile.name,
          },
        };

      const userProfileData = await API
        .graphql(
          graphqlOperation(
            UpdateUserProfile, params,
          ),
        );
      const { data: { updateProfile: newUserProfile } } = userProfileData;
      setUserProfile(newUserProfile);
      console.log(newUserProfile, ' this is NewUserProfile');
      setSuccess({
        header: 'Update successful!',
        content: 'Profile updated successfully.',
      });
      setError(null);
    } catch (e) {
      setSuccess(null);
      setError({
        header: 'Oops...something went wrong',
        content: 'Something went wrong with updating profile, please try again.',
      });
    }
  };
```

**Amplify CLI Version**
You can use `amplify -v` to check the amplify cli version on your system
4.16.1
**To Reproduce**
Steps to reproduce the behavior or terminal output if applicable
1. create new schema
2. create values inside dynamodb table using appsync
3. use update mutation while including the required fields.

**Expected behavior**
The values should update.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: [e.g. Mac/Windows/Ubuntu]
 - Node Version. You can use `node -v` to check the node version on your system

v12.13.1

**Additional context**

**Comment by rush86999** [comment]
I am guessing there's versioning issue?

```
{
    "Version": 4,
    "ResolverConfig": {
        "project": {
            "ConflictHandler": "AUTOMERGE",
            "ConflictDetection": "VERSION"
        },
        "models": {
            ...
            "Profile": {
                "ConflictHandler": "AUTOMERGE",
                "ConflictDetection": "VERSION"
            },
            ...
        }
    }
}
```

**Comment by RossWilliams** [comment]
Can you restate you problem in a different format to better identify the issue? If you can provide the following information needed I’m happy to help.

1. The schema.
2. Any customisations made to resolvers.
2. A GraphQL query or mutation run directly inside the AppSync console. Specify the user permissions used (owner and groups)
3. The result of the query
4. Your expected query result.

**Comment by rush86999** [comment]
1. schema - I will list the relevant parts

```
type User
  @model
  @auth(rules: [
    {allow: groups, groups: ["Managers", "Hosts", "Admin"]},
    {allow: groups, groups: ["public"], operations: [read]},
    {allow: groups, groups: ["Users"], operations: [read]},
    {allow: owner},
    {allow: owner, ownerField: "id"}
  ])
{
  id: ID!
  email: String
  name: String
  profileId: ID!
  weight: String
  profile: Profile @connection(fields: ["profileId"])
  customerIds: [Customer] @connection(keyName: "byUser", fields: ["id"])
  platformCustomerId: String
}

type Profile
@model
@auth(rules: [
  { allow: public, provider: iam, operations: [read] },
  {allow: groups, groups: ["public"], operations: [read]},
  {allow: groups, groups: ["Users"]},
])
{
  id: ID!
  userId: ID!
  name: String!
  aboutMe: String
  profilePicture: String
  following: [UserHost] @connection(keyName: "byUser", fields: ["id"])
  followingCount: String
  user: User @connection(fields: ["userId"])
  stats: Stats @connection(fields: ["userId"])
}
```

2. no customization made to resolvers. Only whatever was autogenerated by the cli

3.  Mutation:
```
export default `mutation ($input: UpdateProfileInput!) {
  updateProfile(input: $input) {
    id
    userId
    name
    aboutMe
    profilePicture
    followingCount
  }
}`;
```
user permission used: owner.

4. result of the query
The initial updated value.  (the VERY FIRST time update was run. Only this value. No new values are updated. 

The field trying to update is profilePicture -  a key that is uploaded on S3 for hosting

After which none of the new updates would work. 

permissions- owner of the table and Users groups used with cognito user pool

5. expected query result

New UPDATED profilePicture key for the most RECENT uploaded picture.

**Comment by rush86999** [comment]
One more thing- there is NO owner field inside Profile table but there is one inside Users table. This is very odd. There should be an owner field for each schema.

**Comment by RossWilliams** [comment]
> One more thing- Inside dynamodb there is NO owner field inside Profile table but there is one inside User table. This is very odd. There should be an owner field for each schema. 

can you edit your above comment to make your schema correct?

**Comment by rush86999** [comment]
I am not sure I understand. "owner" field is autogenerated by the cli. It's not part of the schema? What should I edit?

**Comment by RossWilliams** [comment]
> I am not sure I understand. "owner" field is autogenerated by the cli. It's not part of the schema? What should I edit?

Never mind, I understand what your are trying to say. Profile does not have an owner property in the *generated* schema because there is no owner auth rule on the Profile model.

I’ll take a look at reproducing your issue shortly

**Comment by rush86999** [comment]
oh i see. maybe I can try adding that property and try again but it should work with group ownership of "Users"

**Comment by RossWilliams** [comment]
tl;dr: When you call the "updateProfile" mutation, because you have versioning turned on, you need to pass in the _current_ _version number as part of the update input object. In your react code I can see you aren't doing this.

To test this and show it working properly:
1. Create and push schema (I've removed references to Stats, UserHost, Customer)
```shell
$ yarn global add @aws-amplify/cli@4.16.1

$ amplify init
Enter a name for the project testMutation
? Enter a name for the environment one
? Choose your default editor: Visual Studio Code
? Choose the type of app that you're building javascript
Please tell us about your project
? What javascript framework are you using none
? Source Directory Path:  src
? Distribution Directory Path: dist
? Build Command:  npm run-script build
? Start Command: npm run-script start
Using default provider  awscloudformation

$ amplify add auth
Using service: Cognito, provided by: awscloudformation
 
 The current configured provider is Amazon Cognito. 
 
 Do you want to use the default authentication and security configuration? Default configuration
 Warning: you will not be able to edit these selections. 
 How do you want users to be able to sign in? Username
 Do you want to configure advanced settings? No, I am done.

$ amplify add api
? Please select from one of the below mentioned services: GraphQL
? Provide API name: testmutation
? Choose the default authorization type for the API Amazon Cognito User Pool
Use a Cognito user pool configured as a part of this project.
? Do you want to configure advanced settings for the GraphQL API Yes, I want to make some additional changes.
? Configure additional auth types? Yes
? Choose the additional authorization types you want to configure for the API IAM
? Configure conflict detection? Yes
? Select the default resolution strategy Auto Merge
? Do you have an annotated GraphQL schema? No
? Do you want a guided schema creation? No
? Provide a custom type name Profile

$ amplify push --yes
```

2. In ./amplify/amplify-meta.json, note the values of UserPoolId and AppClientIDWeb
2. create a user for testing
```shell
$ aws cognito-idp admin-create-user --user-pool-id ${UserPoolId} --username testing --temporary-password testing123 --user-attributes Name=email,Value="test@test.com"
$ aws cognito-idp create-group --user-pool-id ${UserPoolId} --group-name Users
$ aws cognito-idp admin-add-user-to-group --user-pool-id ${UserPoolId} --username testing2 --group Users
```

3. navigate to AppSync in your region, https://eu-west-2.console.aws.amazon.com/appsync/ . Select queries on left hand side. Login, use AppClientIDWeb from step 2.

4. execute the following mutations
```gql
mutation create {
  createProfile(input:{id: "456", userId:"testing2", name: "Ross"}) {
    id
    _version
  }
}

mutation update {
  updateProfile(input: {id:"456", profilePicture:"picture1", _version: 1}) {
    id
    profilePicture
    _version
  }
}

mutation update2 {
  updateProfile(input: {id:"456", profilePicture:"picture2", _version: 2}) {
    id
    profilePicture
  }
}
```

5. Notice that profile picture starts undefined, is updated to "picture 1" and then is updated to "picture 2"

**Comment by rush86999** [comment]
I have two roadblocks depending on the solution that would work best.

1. I don't know how to modify to add versioning on client side. I am not sure how the client would be aware of the current version.

2. I really don't need versioning. Is there a way to turn this off? I want to simplify the solution as much as possible so there are no future issues.

**Comment by RossWilliams** [comment]
2. in the cli options for ```amplify add api``` select NO for "? Configure conflict detection?"

1. you can get the current version in the query you do for the Profile, and pass it along to the update method.

**Comment by rush86999** [comment]
oh I see. thanks. I will test this and let you know.

**Comment by ammarkarachi** [comment]
@rush86999 I am closing this issue, do not hesitate to re open this issue if you run into any problems.

**Comment by rush86999** [comment]
@ammarkarachi I am having the problem with delete resolved. Update works now but delete resolver with the same input using _version variable is displaying the same symptoms. Is there anything different I need to do for delete resolved. I decided to go with versioning btw.

**Comment by rush86999** [comment]
i am doing delete with another table.

```
type Appointment
@model
@auth(rules: [
  {allow: public, provider: iam, operations: [read]},
  {allow: groups, groups: ["public"], operations: [read]},
  {allow: groups, groups: ["Users"]},
  {allow: groups, groups: ["Hosts"]},
])
@key(name: "bySchedule", fields: ["scheduleId", "date"], queryField: "appointmentsBySchedule")
@key(name: "byWorkout", fields: ["workoutId", "date"])
{
  id: ID!
  scheduleId: ID!
  hostId: ID
  date: String!
  workoutId: ID!
  duration: String
  workout: Workout @connection(fields: ["workoutId"])
  schedule: Schedule @connection(fields: ["scheduleId"])
  ttl: Int
}
```

graphql 
```
export default `mutation ($input: DeleteAppointmentInput!) {
  deleteAppointment(input: $input) {
    id
    scheduleId
    hostId
    date
    workoutId
    duration
    ttl
    _version
  }
}`;
```

**Comment by rush86999** [comment]
I am passing _version to the input variable for delete mutation.

```
  const deleteAppointment = async () => {
    try {
      setIsCancelling(true);
      console.log(selectedEvent, ' this is selectedEvent inside deleteAppointment');
      const { id, _version } = selectedEvent;

      const deleteAppointmentData = await API
        .graphql(
          graphqlOperation(
            DeleteAppointment, {
              input: {
                id,
                _version,
              },
            },
          ),
        );

      const {
        data: {
          deleteAppointment: removedAppointment,
        },
      } = deleteAppointmentData;

      console.log(removedAppointment, ' this is deleted appointment data');

      //  reset selected state and get new schedule

      setIsCancelling(false);
      setIsConfirmCancel(false);
      const bool = removeAppointmentFromEvents();

      if (bool) {
        setSuccess({
          header: 'Success!',
          content: 'Event successfully removed',
        });
      } else {
        setError('Oops...something went wrong, try refreshing the app');
      }
    } catch (e) {
      console.error('unable to delete slots', e);
      setIsCancelling(false);
      setError(`Oops...something went wrong: ${e}`);
    }
  };
```

**Comment by rush86999** [comment]
The operation is being performed by a user part of "Hosts" group using cognito authentication

**Comment by rush86999** [comment]
One other issue is why are the docs mentioning expectedVersion when no such field exists? it's _version. Please update the docs to reflect this and also delete resolver. Is it possible to get an ETA on this as I might decide not to use version control. Also if I already have a project setup can I use "amplify update api" to remove conflict resolution as I don't want to create a whole new project and start from scratch.

**Comment by undefobj** [comment]
@rush86999 Looking at your issue it appears that you have enabled Conflict Resolution on your API which is designed to work with [Amplify DataStore](https://docs.amplify.aws/lib/datastore/getting-started/q/platform/js). The way this is designed to work is that the object version is tracked at the service and in the client as well, and when it does a create/update operation the current version is sent transparently along with the request. If a conflict is detected it will enter one of the conflict resolution strategies such as Auto Merge which is the default configuration. You can learn more about this here:

https://docs.amplify.aws/lib/datastore/how-it-works/q/platform/js

It looks like you are intending to use your GraphQL API with the regular API category and not DataStore. If thats the case then I would recommend creating an API without sync enabled (e.g. do not use `amplify-app` npx script and instead follow the `amplify add api` standard flow). If you'd like to try turning off Conflict Resolution for your existing API you can do this with the latest CLI like so:

```
$ amplify update api
? Please select from one of the below mentioned services: GraphQL
? Select from the options below Walkthrough all configurations
? Choose the default authorization type for the API API key
? Enter a description for the API key: 
? After how many days from now the API key should expire (1-365): 7
? Do you want to configure advanced settings for the GraphQL API Yes, I want to 
make some additional changes.
? Configure additional auth types? No
? Configure conflict detection? No
```

**Comment by undefobj** [comment]
Alternatively in your conf file: https://github.com/aws-amplify/amplify-cli/issues/3726#issuecomment-601969958

Remove the `ResolverConfig` key:

```
{
    "Version": 4,
}
```

Then do an `ampllify push`.

**Comment by rush86999** [comment]
Thanks for the info. I removed conflict resultion and the API works. However, I m unable to have to have unauthenticated access to many of the dynamic links created via react router in create app. Amplify seems to initiate the authorized login process as if a user landed from a third party social authentication authorization. It tries to log me in when this is not the intention. I don't think Amplify figured out a nice way to add social login without messing with unauthenticated access so I have put the project on hold.

**Comment by kldeb** [comment]
@undefobj what if we want to use the DataStore in the future? Will adding DataStore later cause any issues? 

Seems to me that requiring _version field on updates would be straightforward (worked for us), is this difficult to support?

**Comment by kldeb** [comment]
What's the migration path from non DataStore to DataStore? 

There needs to be front end and back end changes...

Create new environment to make changes then merge those changes into the main branch?
